File: Backend/src/main/java/com/example/backend/controller/ProductController.java
package com.example.backend.controller;

import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.example.backend.dto.ApiResponse;
import com.example.backend.entities.Product;
import com.example.backend.service.ProductService;

@RestController
@RequestMapping("/products")
@CrossOrigin
public class ProductController {

    @Autowired
    private ProductService productService;

    @GetMapping
    public ApiResponse getAllProducts(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "40") int size,
            @RequestParam(defaultValue = "id") String sortBy,
            @RequestParam(defaultValue = "desc") String order
    ) {
        try {
            return productService.getAllProducts(page, size, sortBy, order); 
        } catch (Exception e) {
            return ApiResponse.error("Lỗi: " + e.getMessage());
        }
    }

    @GetMapping("/{id}")
    public ApiResponse getProductDetail(@PathVariable Long id) {
        try {
            return productService.getProductDetail(id);
        } catch (Exception e) {
            return ApiResponse.error("Lỗi: " + e.getMessage());
        }
    }
    
    @PostMapping
    public ApiResponse createProduct(@RequestBody Product product) {
         try {
            return productService.createProduct(product);
        } catch (Exception e) {
            return ApiResponse.error("Lỗi: " + e.getMessage());
        }
    }

    @GetMapping("/search/{keyword}")
    public ApiResponse searchProducts(@PathVariable String keyword) {
        try {
            return productService.searchProducts(keyword);
        } catch (Exception e) {
            return ApiResponse.error("Lỗi tìm kiếm: " + e.getMessage());
        }
    }

    @GetMapping("/category/{categories}")
    public ApiResponse filterProducts(
            @PathVariable List<String> categories,
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "20") int size,
            @RequestParam(defaultValue = "id") String sortBy,
            @RequestParam(defaultValue = "desc") String order
    ) {
        try {
            return productService.filterProducts(categories, page, size, sortBy, order);
        } catch (Exception e) {
            return ApiResponse.error("Lỗi lọc: " + e.getMessage());
        }
    }

    @GetMapping("/best-sellers")
    public ApiResponse getBestSellers() {
        try {
            return productService.getTop10BestSellers();
        } catch (Exception e) {
            return ApiResponse.error("Lỗi: " + e.getMessage());
        }
    }
}

File: Frontend/src/controllers/product.controller.ts
import { Product, ProductDetail } from "@/src/models/product.model";
import { productService } from "@/src/services/product.service";

type ProductResult = 
  | { ok: true; data: Product[] }
  | { ok: false; message: string };

type ProductDetailResult = 
  | { ok: true; data: ProductDetail }
  | { ok: false; message: string };

export const productController = {
  async getList(page = 0, size = 20, sortBy = 'id', order = 'desc'): Promise<ProductResult> {
    try {
      const res = await productService.getAllProducts(page, size, sortBy, order);
      
      if (res.success && res.data) {
        // console.log(res.data);
        return { ok: true, data: res.data.content }; 
      }
      
      return { ok: false, message: res.message || "Không tải được danh sách" };
    } catch (e: any) {
      return { ok: false, message: e.message || "Lỗi kết nối server" };
    }
  },

  async search(keyword: string): Promise<ProductResult> {
    try {
      const res = await productService.searchProducts(keyword);
      if (res.success && res.data) return { ok: true, data: res.data };
      return { ok: false, message: res.message || "Không tìm thấy sản phẩm" };
    } catch (e: any) {
      return { ok: false, message: e.message || "Lỗi kết nối" };
    }
  },

  async filter(categories: string[], page = 0, size = 20, sortBy = 'id', order = 'desc'): Promise<ProductResult> {
    try {
      const res = await productService.filterByCategory(categories, page, size, sortBy, order);
      
      if (res.success && res.data) {
        const products = res.data.content || res.data; 
        return { ok: true, data: products };
      }
      return { ok: false, message: res.message || "Lỗi lọc sản phẩm" };
    } catch (e: any) {
      return { ok: false, message: e.message || "Lỗi kết nối" };
    }
  },

  async get(id: number): Promise<ProductDetailResult> {
    try {
      const res = await productService.getProductDetail(id);
      
      if (res.success && res.data) {
        return { ok: true, data: res.data };
      }
      
      return { ok: false, message: res.message || "Không tải được thông tin sản phẩm" };
    } catch (e: any) {
      return { ok: false, message: e.message || "Lỗi kết nối server" };
    }
  },

  async getBestSellers(): Promise<ProductResult> {
    try {
      const res = await productService.getBestSellers();
      if (res.success && res.data) {
        return { ok: true, data: res.data }; 
      }
      return { ok: false, message: res.message || "Không tải được sản phẩm bán chạy" };
    } catch (e: any) {
      return { ok: false, message: e.message || "Lỗi kết nối" };
    }
  },

  async getSimilar(id: number): Promise<{ ok: true; data: Product[] } | { ok: false; message: string }> {
    try {
      const res = await productService.getSimilarProducts(id);
      if (res.success && res.data) return { ok: true, data: res.data };
      return { ok: false, message: res.message || "Lỗi tải sản phẩm tương tự" };
    } catch (e: any) {
      return { ok: false, message: e.message || "Lỗi kết nối" };
    }
  },
};